includes/core/server/request/trait-use-test-mode-only-when-dev-mode.php

```diff
@@ -0,0 +1,31 @@
+<?php
+/**
+ * Trait file for WCPay\Core\Server\Request\Request_Test_Mode_Only_For_Dev_Mode.
+ *
+ * @package WooCommerce Payments
+ */
+
+namespace WCPay\Core\Server\Request;
+
+use WC_Payments;
+use WC_Payments_API_Client;
+use WC_Payments_Http_Interface;
+use WCPay\Core\Exceptions\Server\Request\Invalid_Request_Parameter_Exception;
+
+/**
+ * Trait for requests that we only send a test mode request if in dev mode.
+ */
+trait Use_Test_Mode_Only_When_Dev_Mode {
+	/**
+	 * Class constructor.
+	 *
+	 * @param  WC_Payments_API_Client     $api_client  Api client.
+	 * @param  WC_Payments_Http_Interface $http_interface  Http interface.
+	 *
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	public function __construct( WC_Payments_API_Client $api_client, WC_Payments_Http_Interface $http_interface ) {
+		parent::__construct( $api_client, $http_interface );
+		$this->set_param( 'test_mode', WC_Payments::mode()->is_dev() );
```

Comment:
I would avoid this approach. There is `DEFAULT_PARAMS` inside the abstract request class that could be used for these cases. Overriding the constructor makes things difficult because you will have to maintain the constructor param injection in multiple places.

Reply:
The problem is that the dev mode value can be retrieved via an expression only, which does not support for const like `DEFAULT_PARAMS`.

![Markup on 2023-04-03 at 17:34:24](https://user-images.githubusercontent.com/10045087/229485942-3a681b5b-d0b0-420b-8478-ec8bf8ac886e.png)

> Overriding the constructor makes things difficult because you will have to maintain the constructor param injection in multiple places.

Not really, as in our usage, we use `create()`, which hides the constructor initialization. I myself don't like it just because it hides the constructor in a trait, which makes things hard to read. However, by introducing a trait, we can easily signify which classes have the same attribute.

Reply:
Rather than creating a new trait that overrides the constructor, a better approach could be to add a specific function to the existing trait you added that handles the dev mode requirement in intent requests. This way, we can build upon our previous work and improve the trait's functionality without creating unnecessary complexity in the code.

For example, you can put that `test_mode` is required and inside trait add something like this:

```
public function add_test_mode_paramameter() {
      $this->set_param( 'test_mode', WC_Payments::mode()->is_dev() );
}
```

and call this function when you call other's setters. Or you could create `set_test_mode` function and fill it manually.

Reply:
@zmaglica - awesome feedback. I've addressed it here 6604de1e3f6aa6e6388811fc652bab5d4469094f 

I am marking this as resolved now. In case, you have any feedback regarding the new approach, feel free to leave inline comments.